imap_processing/utils.py

```diff
@@ -254,3 +255,86 @@ def update_epoch_to_datetime(dataset: xr.Dataset) -> xr.Dataset:
+    packet definition, the ``derived_value`` will be the converted value.
+    The dimension of the dataset will be the time field in J2000 nanoseconds.
+
+    Parameters
+    ----------
+    packet_file : str
+        Path to data packet path with filename.
+    xtce_packet_definition : str
+        Path to XTCE file with filename.
+
+    Returns
+    -------
+    datasets : dict
+        Mapping from apid to xarray dataset, one dataset per apid.
+    """
+    # Set up containers to store our data
+    # We are getting a packet file that may contain multiple apids
+    # Each apid has consistent data fields, so we want to create a
+    # dataset per apid.
+    # {apid1: dataset1, apid2: dataset2, ...}
+    data_dict: dict[int, dict] = dict()
+    attribute_dict: dict[int, dict] = dict()
+
+    # Set up the parser from the input packet definition
+    packet_definition = xtcedef.XtcePacketDefinition(xtce_packet_definition)
+    packet_parser = parser.PacketParser(packet_definition)
+
+    with open(packet_file, "rb") as binary_data:
+        packet_generator = packet_parser.generator(binary_data)
+        for packet in packet_generator:
+            apid = packet.header["PKT_APID"].raw_value
+            if apid not in data_dict:
+                # This is the first packet for this APID
+                data_dict[apid] = collections.defaultdict(list)
+                attribute_dict[apid] = collections.defaultdict(dict)
+
+                # Fill in the attributes
+                for key, value in packet.data.items():
+                    attribute_dict[apid][key]["short_description"] = (
+                        value.short_description
+                    )
+                    attribute_dict[apid][key]["long_description"] = (
+                        value.long_description
+                    )
+                    # TODO: Additional info from XTCE like units
+
+            # TODO: Include the header too? (packet.header | packet.data)
+            for key, value in packet.data.items():
+                # TODO: Do we want derived_value or raw_value?
+                # Right now we use a derived value is there is one, otherwise raw value
+                # Maybe we want these to be separated into l1a (raw) and l1b (derived)?
+                data_dict[apid][key].append(value.derived_value or value.raw_value)
+
+    dataset_by_apid = {}
+    # Convert each apid's data to an xarray dataset
+    for apid, data in data_dict.items():
+        # The time key is always the first key in the data dictionary on IMAP
+        time_key = next(iter(data.keys()))
+        # Convert to J2000 time and use that as our primary dimension
+        time_data = met_to_j2000ns(data[time_key])
+        ds = xr.Dataset(
+            {key.lower(): ("epoch", val) for key, val in data.items()},
+            coords={"epoch": time_data},
+        )
+        for key, value in attribute_dict[apid].items():
+            ds[key.lower()].attrs.update(value)
```

Comment:
Is the assumption here that all of the required CDF attributes will be defined in the xtce? or would there be another round of updating the variable attributes in the dataset?

Reply:
No, I view this as very much a "just give me an xarray dataset". Then I'll start appending metadata, adding more stuff to it etc... We could leave this out if it'd be easier to understand and use later, but I figured it might be useful for housekeeping datasets.

Reply:
Ok, great. I wonder if it is even useful to include the attributes defined in XTCE or if we should just centralize on using the YAML for all attributes.

Reply:
I ended up removing the attributes and short/long descriptions. It simplified things, so I think makes sense for now.